### PR TITLE
[PSDK-348] Trade Verb Support w/ MPC Server-Signer

### PR DESCRIPTION
**Pull request by John-peterson-coinbase**
### What changed? Why?
- Support for trade with MPC Server-Signer
- `CreateTradeOptions` type
-  Creating a trade promise now waits for the trade to land on chain to be consistent with transfers

#### Qualified Impact

This may impact all trade flows. Fix-forward, if necessary.

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

File: src/coinbase/address/wallet_address.ts [hunk 1]
```diff
@@ -244,25 +251,53 @@ export class WalletAddress extends Address {
   /**
    * Trades the given amount of the given Asset for another Asset. Only same-network Trades are supported.
    *
-   * @param amount - The amount of the Asset to send.
-   * @param fromAssetId - The ID of the Asset to trade from.
-   * @param toAssetId - The ID of the Asset to trade to.
+   * @param options = The options to create the Trade.
+   * @param options.amount - The amount of the Asset to send.
```
Comment (jazz-cb) [inline src/coinbase/address/wallet_address.ts:255]: The amount of the fromAsset to trade

File: src/coinbase/address/wallet_address.ts [hunk 2]
```diff
@@ -244,25 +251,53 @@ export class WalletAddress extends Address {
...
+   * @param options.fromAssetId - The ID of the Asset to trade from.
+   * @param options.toAssetId - The ID of the Asset to trade to.
+   * @param options.timeoutSeconds - The maximum amount of time to wait for the Trade to complete, in seconds.
+   * @param options.intervalSeconds - The interval at which to poll the Network for Trade status, in seconds.
    * @returns The Trade object.
-   * @throws {Error} If the private key is not loaded, or if the asset IDs are unsupported, or if there are insufficient funds.
+   * @throws {APIError} if the API request to create a Trade fails.
```
Comment (jazz-cb) [inline src/coinbase/address/wallet_address.ts:261]: can this be `@throws {APIError} if the API request to create a Trade or broadcast fails.`

File: src/coinbase/address/wallet_address.ts [hunk 3]
```diff
@@ -244,25 +251,53 @@ export class WalletAddress extends Address {
...
+    const startTime = Date.now();
+    while (Date.now() - startTime < timeoutSeconds * 1000) {
+      await trade.reload();
+      const status = trade.getStatus();
+      if (status === TransactionStatus.COMPLETE || status === TransactionStatus.FAILED) {
+        return trade;
+      }
+      await delay(intervalSeconds);
+    }
+    throw new Error("Trade timed out");
```
Comment (jazz-cb) [inline src/coinbase/address/wallet_address.ts:300]: else we should return trade?

Comment (John-peterson-coinbase) [reply src/coinbase/address/wallet_address.ts:300]: +1. Lets follow up with this handling on both trades and transfers.

**Review by jazz-cb — APPROVED** [review]

**Pull request closed without merging**